### ISSUE TITLE
Logging at training/validation batch end

### ISSUE THREAD

**Issue by zhutmost** [PyTorchLightning/pytorch-lightning#4985]
## 🚀 Feature
Currently it seems that PL cannot log metrics to progress bar in `training/validation_step_end` with `self.log()`. 
I try to `self.log_dict({...})` in these functions, but the progress bar only prints loss without other metrics. I want to log metrics in `<mode>_step_end` just like what I can do in `<mode>_step`

### Motivation
Currently PL cannot log metrics in `training/validation_step_end` of a `LightningModule`. These two functions are necessary in DataParallel training.

### Pitch
log metrics in `<mode>_step_end` just like what I can do in `<mode>_step`.

### Alternatives
None

### Additional context
I located some related codes in the PL repository, and i want it can help you the developers. 
The following screenshot is line 557-583 of `training_loop.py`.
<img width="850" alt="image" src="https://user-images.githubusercontent.com/20497382/101258409-4b50ba00-375d-11eb-8a58-e4323ff1a00d.png">
At the line 561, the function `self.run_training_batch(batch, batch_idx, dataloader_idx)` returns `batch_output`, which contains the logged metrics in `training_step`. And then the training loop passes these metrics to another function `self.trainer.logger_connector.log_train_step_metrics(batch_output)` at Line 582.

However, on the other hand, the function `self.on_train_batch_end(epoch_output, epoch_end_outputs, batch, batch_idx, dataloader_idx)` doesn't return anything. Is this the reason why I cannot log the metrics in `training_step_end`?

**Comment by zhutmost** [comment]
If this bug cannot be fixed soon, is there any temporary workaround?
